res/controllers/Pioneer-DDJ-400-script.js

```diff
@@ -0,0 +1,741 @@
+//
+//  Implemented (as per manufacturer's manual):
+//      * Mixer Section (Faders, EQ, Filter, Gain, Cue)
+//      * Browsing and loading + Waveform zoom (shift)
+//      * Jogwheels, Scratching, Bending, Loop adjust
+//      * Cycle Temporange
+//      * Beat Sync
+//      * Hot Cue Mode
+//      * Beat Loop Mode
+//      * Beat Jump Mode
+//      * Sampler Mode
+//
+//  Custom (Mixxx specific mappings):
+//      * BeatFX: Assigned Effect Unit 1
+//                < LEFT selects EFFECT1
+//                > RIGHT selects EFFECT2
+//                v FX_SELECT selects EFFECT3.
+//                v again to get back to wet/dry mix
+//                ON/OFF toggles selected effect slot
+//                SHIFT + ON/OFF disables all three effect slots.
+//                SHIFT + < selects previous effect
+//                SHIFT + > selects next effect
+//
+//      * 32 beat jump forward & back (Shift + </> CUE/LOOP CALL arrows)
+//      * Toggle quantize (Shift + channel cue)
+//
+//  Not implemented (after discussion and trial attempts):
+//      * Loop Section:
+//        * -4BEAT auto loop (hacky---prefer a clean way to set a 4 beat loop
+//                            from a previous position on long press)
+//
+//        * CUE/LOOP CALL - memory & delete (complex and not useful. Hot cues are sufficient)
+//
+//      * Secondary pad modes (trial attempts complex and too experimental)
+//        * Keyboard mode
+//        * Pad FX1
+//        * Pad FX2
+//        * Keyshift mode
+
+var PioneerDDJ400 = {};
+
+PioneerDDJ400.lights = {
+    beatFx: {
+        status: 0x94,
+        data1: 0x47,
+    },
+    deck1: {
+        vuMeter: {
+            status: 0xB0,
+            data1: 0x02,
+        },
+        playPause: {
+            status: 0x90,
+            data1: 0x0B,
+        },
+        shiftPlayPause: {
+            status: 0x90,
+            data1: 0x47,
+        },
+        cue: {
+            status: 0x90,
+            data1: 0x0C,
+        },
+        shiftCue: {
+            status: 0x90,
+            data1: 0x48,
+        },
+    },
+    deck2: {
+        vuMeter: {
+            status: 0xB0,
+            data1: 0x02,
+        },
+        playPause: {
+            status: 0x91,
+            data1: 0x0B,
+        },
+        shiftPlayPause: {
+            status: 0x91,
+            data1: 0x47,
+        },
+        cue: {
+            status: 0x91,
+            data1: 0x0C,
+        },
+        shiftCue: {
+            status: 0x91,
+            data1: 0x48,
+        },
+    },
+};
+
+// Store timer IDs
+PioneerDDJ400.timers = {};
+
+// Stores padmode state and encapsulates LED logic
+PioneerDDJ400.performancePads = {
+    modes: {
+        HOTCUE: 0x1B,
+        BEATLOOP: 0x6D,
+        BEATJUMP: 0x20,
+        SAMPLER: 0x22,
+        KEYBOARD: 0x69,
+        PADFX1: 0x1E,
+        PADFX2: 0x6B,
+        KEYSHIFT: 0x6F
+    },
+
+    // Start in hotcue mode
+    state: [
+        0x1B, // Channel 1
```

Comment:
Are you able to reference `modes.HOTCUE` here? Not sure if JS has a nice way to do that in this case.

Reply:
Agreed it's not ideal. Unfortunately the object has to be initialized for the `modes` hash to be referenced. Could be moved into a global?

Reply:
Nah, leave it as it is.